src/main/java/com/puppycrawl/tools/checkstyle/FileStatefulModule.java

```diff
@@ -0,0 +1,48 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2017 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * This annotation means that the module contains file-related context and therefore
+ * cannot be used from the separate threads at the same time.
+ * This annotation should be used when a module holds a thread-unsafe state.
+ * It is guaranteed that the whole file will be processed inside the same thread.
+ * It is guaranteed that the whole file will be processed with the same module instance.
+ * It is guaranteed that each module instance processes only one file at the same time.
+ * It is guaranteed that all module clones will have equal (but not the same) configuration.
```

Comment:
> clones

we we still do clones? I though we already discussed that this is not requied.

> equal (but not the same) configuration

does it really matter  to mention  ? Checks arguments should be (usually) immutable.

Reply:
> we we still do clones? I though we already discussed that this is not requied.

We are not cloning them, they are created from the XML configuration, when needed. How they should be named?

Reply:
> Checks arguments should be (usually) immutable.

Sure, but there are some properties of type `String[]`. By `equal (but not the same) configuration` I mean, that two check instances will not hold the same array. Should I remove this note?

Reply:
> How they should be named?

new instances.

> Should I remove this note?

please extend that line and mention that there are arrays.
But to my mind it is Check responsibility to make a copy of user defined properties to be not affected by mutual values.

Reply:
@soon , is it fixed ?

Reply:
Fixed